src/DotVVM.Framework/Binding/DotvvmProperty.cs

```diff
@@ -238,29 +238,24 @@ public static void InitializeProperty(DotvvmProperty property, ICustomAttributeP
         }
 
-        public static IEnumerable<DotvvmProperty> GetVirtualProperties(Type controlType)
-            => from p in controlType.GetProperties(BindingFlags.Public | BindingFlags.Instance)
-               where !registeredProperties.ContainsKey(p.DeclaringType!.FullName + "." + p.Name)
-               let markupOptions = GetVirtualPropertyMarkupOptions(p)
-               where markupOptions != null
-               where p.GetCustomAttribute<PropertyGroupAttribute>() == null
-               where markupOptions.MappingMode != MappingMode.Exclude
-               select new DotvvmProperty {
-                   DeclaringType = controlType,
-                   IsValueInherited = false,
-                   MarkupOptions = markupOptions,
-                   Name = p.Name,
-                   PropertyInfo = p,
-                   PropertyType = p.PropertyType,
-                   IsVirtual = true
-               };
-
-        private static MarkupOptionsAttribute? GetVirtualPropertyMarkupOptions(PropertyInfo p)
+        public static void CheckAllPropertiesAreRegistered(Type controlType)
         {
-            var mo = p.GetCustomAttribute<MarkupOptionsAttribute>();
-            if (mo == null) return null;
-            mo.AllowBinding = false;
-            return mo;
+            var properties =
+               (from p in controlType.GetProperties(BindingFlags.Public | BindingFlags.Instance)
+                where !registeredProperties.ContainsKey(p.DeclaringType!.FullName + "." + p.Name)
+                where p.GetCustomAttribute<PropertyGroupAttribute>() == null
+                let markupOptions = p.GetCustomAttribute<MarkupOptionsAttribute>()
+                where markupOptions != null && markupOptions.MappingMode != MappingMode.Exclude
+                select p).ToArray();
+
+            if (properties.Any())
+            {
+                var controlHasOtherProperties = registeredProperties.Values.Any(p => p.DeclaringType == controlType);
+                var explicitLoadingHelp =
+                    !controlHasOtherProperties ? $" The control does not have any other properties registered, it could indicate that DotVVM did not register properties from this assembly - it is common with ExplicitAssemblyLoading enabled, try to register {controlType.Assembly} into config.Markup.Assemblies." : "";
+                var deprecationHelp = " DotVVM version <= 3.x did support this, but this feature was removed as it lead to many issues. Please register the property using DotvvmProperty.Register. If you find this annoyingly verbose, you could use control capabilities instead (using DotvvmCapabilityProperty.Register).";
+                throw new NotSupportedException($"Control {controlType} has properties which are not registered as DotvvmProperty but have a MarkupOptionsAttribute: {string.Join(", ", properties.Select(p => p.Name))}." + explicitLoadingHelp + deprecationHelp);
```

Comment:
Since this exception message is rather long, I have a suggestions to improve its typography:

* Use quotation marks around arguments to be consistent with most of our other exception messages (`'{controlType}'`).
* Either put a comma before "which", or use "that".
* "a DotvvmProperty".
* "...any other properties registered. It could indicate..."
* I think we can omit "- it is common with ExplicitAssemblyLoading enabled" for brevity.
* "DotVVM version <= 3.x" -> "DotVVM versions prior to 3.0"
* "...using DotvvmProperty.Register or use control capabilities instead."

Reply:
"DotVVM versions prior to 3.0" - that would be <

> I think we can omit "- it is common with ExplicitAssemblyLoading enabled" for brevity.

I think it's useful information, this error is quite unlikely to occur without this option and adding to Assemblies will not help if ExplicitAssemblyLoading is disabled.